opm/autodiff/WellStateFullyImplicitBlackoil.hpp

```diff
@@ -282,6 +293,196 @@ namespace Opm
+                        if (pu.phase_used[Gas]) {
+                            const int gaspos = pu.phase_pos[Gas];
+                            // scale the phase rates for Gas to avoid too bad initial guess for gas fraction
+                            // it will probably benefit the standard well too, while it needs to be justified
+                            // TODO: to see if this strategy can benefit StandardWell too
+                            // TODO: it might cause big problem for gas rate control or if there is a gas rate limit
+                            // maybe the best way is to initialize the fractions first then get the rates
+                            for (int perf = 0; perf < nperf; perf++) {
+                                const int perf_pos = start_perf + perf;
+                                perfPhaseRates()[np * perf_pos + gaspos] *= 100.;
+                            }
+                        }
+
+                        const std::vector<double> perforation_rates(perfPhaseRates().begin() + np * start_perf,
+                                                                    perfPhaseRates().begin() + np * start_perf_next_well); // the perforation rates for this well
+                        std::vector<double> segment_rates;
+                        calculateSegmentRates(segment_inlets, segment_perforations, perforation_rates, np, 0 /* top segment */, segment_rates);
+                        std::copy(segment_rates.begin(), segment_rates.end(), std::back_inserter(segrates_));
+                    }
+
+                    // for the segment pressure, the segment pressure is the same with the first perforation
+                    // if there is no perforation associated with this segment, it uses the pressure from the outlet segment
+                    // which requres the ordering is successful
+                    // TODO: maybe also check the relation with the cell?
+                    // Not sure what is the best way to handle the initialization, hopefully, the bad initialization can be
+                    // improved during the solveWellEq process
+                    {
+                        // top segment is always the first one, and its pressure is the well bhp
+                        segpress_.push_back(bhp()[w]);
+                        const int top_segment = top_segment_loc_[w];
+                        const int start_perf = wells->well_connpos[w];
+                        for (int seg = 1; seg < nseg; ++seg) {
+                            if ( !segment_perforations[seg].empty() ) {
+                                const int first_perf = segment_perforations[seg][0];
+                                segpress_.push_back(perfPress()[start_perf + first_perf]);
+                            } else {
+                                // segpress_.push_back(bhp); // may not be a good decision
+                                // using the outlet segment pressure // it needs the ordering is correct
```

Comment:
Perhaps add an assert() to make sure that the outlet segment pressure has already been computed? Alternatively use segpress_.at(...) instead of segpress_[...] on line 416 to make it throw if invalid.

Reply:
The ordering makes the outlet segment always have a lower index than the segment. I think it is okay.